etcd/cluster-etcd-operator.md

```diff
@@ -0,0 +1,239 @@
+---
+title: cluster-etcd-operator
+authors:
+  - "@hexfusion"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@deads2k"
+  - "@sttts"
+approvers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+  - "@sttts"
+creation-date: 2019-10-1
+last-updated: 2019-10-2
+status: implementable
+see-also: 
+replaces:
+superseded-by:
+---
+
+# Cluster etcd Operator
+
+cluster-etcd-operator (CEO) is an operator that handles the scaling of etcd during cluster bootstrap and regular operation and provisioning etcd dependencies such as TLS certificates.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Open Questions [optional]
+
+TODO
+
+## Summary
+
+TODO
+
+## Motivation
+
+Having an operator manage the etcd cluster members requires the ability to scale up and scale down the etcd membership separate from the actual node itself. Meaning if you were to scale down from 5 to 4 that does not necessarily mean that you must remove the node from the cluster to achieve this. A good example would be replacing a failed member. In this case we need to remove the failed member (scale down to 4) and re-add (scale back to 5) no need for new node in this case. Scaling etcd requires a balance of proper configuration and timing to minimize stress to cluster.
+
+The most dangerous scaling is from a single-member cluster (bootstrap) to a two-member cluster. This is dangerous because as soon as the etcd API is notified of the new member the cluster loses quorum until that new member starts and joins the cluster. To account for these dangers we utilize the init containers to provide structured gates. These gates allow for data to communicate from the static pod init container and the operator using a ConfigMap.
+
+### Goals
+
+1. Deploy etcd instance on the bootstrap node to fast track CVO deployment.
+2. Scaling up/down of etcd peer membership.
+3. Replace failed member assuming quorate cluster.
+4. Automation of disaster recovery tasks (human interaction required)
+
+### Non-Goals
+
+1. Automated disaster recovery of a cluster that has lost quorum.
+
+## Proposal
+
+API:
+
+```
+
+type EtcdScaling struct {
+    Metadata *metav1.ObjectMeta `json:"metadata,omitempty"`
```

Comment:
Can we s/Metadata *metav1.ObjectMeta/Name string/? I was looking at the code and all we are using is the name?

Reply:
Yeah this data is redundant/unused I updated the API to reflect current design.